### DISCUSSION TITLE
How does Eclipse-based editor work

### DISCUSSION THREAD

**Discussion by Avaritia1** [eclipse-glsp/glsp#1007] [Q&A]
Hi! I am trying to implement a EMF.cloud based GLSP project on Eclipse IDE. The idea is to combine the backend part with both server components from [this project](https://github.com/eclipse-emfcloud/modelserver-glsp-integration/tree/main/project-templates/modelserver-glspjava-emf-theia/server) and the frontend integration part from [this project](https://github.com/eclipse-glsp/glsp-examples/tree/master/project-templates/java-emf-eclipse/glsp-client).

The goal is to be able to open and interact with two instances shown in the image below locally. For example: After deleting a node in one instance, the node in another instance is also deleted.

However, I have a question about this Eclipse-based editor. When I open the [glsp-client file](https://github.com/eclipse-glsp/glsp-examples/tree/master/project-templates/java-emf-eclipse/glsp-client) in Eclipse, it shows 3 Java files instead of the ts files in the glsp-client file. I wonder if these Java files are generated from ts files and how these files work. Thanks in advance!
![image](https://github.com/eclipse-glsp/glsp/assets/80332537/37e00376-eef0-4cd6-adcd-847369de02d1)

**Comment by ndoschek** [comment]
Hi @Avaritia1,
great to see you already made progress on your GLSP Eclipse endeavour! :tada: 
I will have a look into it in the next days and will come back to you!
BR, Nina

**Reply by Avaritia1** [reply]
Hi! I wonder if there are any updates about my question :) Thanks!

**Reply by ndoschek** [reply]
Hi @Avaritia1,
sorry for the delay!

Basically, the `*.tasklist` file contains the semantic model (the actual `Tasklist`model instance) and the `*.notation` model contains the notation model, which holds the information about the positioning of the semantic `Task` elements.
The diagram editor combines the information of both models to generate a graphic model which is then used to render the diagram.

You could also open them with a text editor to have a look at the textual representation of the models (indicated by a simple file icon).

However, I'm not entirely sure what you mean by "3 Java files instead of the ts files in the glsp-client file", could you please elaborate on that? Thanks!

HTH and please let me know if there are any other open questions :)
Best, Nina

**Reply by Avaritia1** [reply]
Hi @ndoschek, thanks for the information! Yes, I can elaborate on it. What I said about Java files are Activator.java, TaskListEclipseDiagramModule.java and TaskListServerManager in [org.eclipse.glsp.example.javaemf.editor](https://github.com/eclipse-glsp/glsp-examples/tree/master/project-templates/java-emf-eclipse/glsp-server/org.eclipse.glsp.example.javaemf.editor/src/org/eclipse/glsp/example/javaemf/editor). But when I open the [server part of Eclipse EMF.cloud - Project Template](https://github.com/eclipse-emfcloud/modelserver-glsp-integration/tree/main/project-templates/modelserver-glspjava-emf-theia/server) in VS Code, it doesn't show this part of code. I wonder if this part of the code is replaced by other parts, or is there another way to do it?


![image](https://github.com/eclipse-glsp/glsp/assets/80332537/be239c50-a1ea-4c11-997a-943b0b9ed23f)
![image](https://github.com/eclipse-glsp/glsp/assets/80332537/86539178-a9b0-4c1e-b2e5-87ee424c54cf)

**Reply by ndoschek** [reply]
Hi @Avaritia1,
thank you for clearing that up!

Yes, you are already on the right track, basically on the backend you will need to additionally add the `org.eclipse.glsp.example.javaemf.editor` project  from the [eclipse-integration example](https://github.com/eclipse-glsp/glsp-examples/tree/master/project-templates/java-emf-eclipse/glsp-server/org.eclipse.glsp.example.javaemf.editor).

I'll try to join the dots between the components and their responsibilities:
1. The main package.json build script will first build the client packages and then the server projects.
1. Next, the client packages  (ts files) are compiled and on success copied to the server project `org.eclipse.glsp.example.javaemf.editor/diagram` ([see here for details](https://github.com/eclipse-glsp/glsp-examples/blob/c79f688cfa9227be82627753db6fae5e838dcd0b/project-templates/java-emf-eclipse/glsp-client/package.json#L26)). This ensures that the files are ready to be rendered in the built-in webbrowser of Eclipse.
1. Next, all server projects are built (please see remark below).
1. Then, the main runnable (`org.eclipse.glsp.example.javaemf.editor/TaskListEditor.launch`) is started, which opens a runtime instance of Eclipse and allows to open and edit `*.tasklist` diagrams.

Remark: There will be some more adaptations needed on the backend:
- The `project-templates/java-emf-eclipse` uses a target platform to retrieve the dependencies, while the `project-templates/modelserver-glspjava-emf-theia` retrieves all dependencies via m2. I'll suggest you would switch those two projects to plugin projects as well (so that they retrieve their dependencies from the target platform build as well).

To summarize, the project structure should look coarsely like this:
- client
  - tasklist-glsp
  - tasklist-eclipse
- server
  - glspjavaserver
  - modelserver
  - editor
  - target

HTH and best regards,
Nina
